### DISCUSSION TITLE
Multiple FlyleafHost Matrix Switching

### DISCUSSION THREAD

**Discussion by webclient163** [SuRGeoNix/Flyleaf#287] [Q&A]
When I use multiple FlyleafHost to play 16 videos in one window, and then switch to playing only 4 of them (the other 12 videos are not displayed on the screen temporarily, in order to switch quickly, these 12 videos are not turned off, but just not displayed on the screen), if the FlyleafHost.Dispose() method is called to free FlyleafHost, the memory will continue to grow and quickly cause memory exhaustion, if Dispose() is not used) The method is that the memory will not be freed, and multiple switches will cause the memory to gradually run out!

**Comment by SuRGeoNix** [comment]
@webclient163 FlyleafHost's disposal is under discussion, it currently considers that it will be disposed when the Owner window will close. However, the issue that you mention probably is the FlyleafHost.Player and not the FlyleafHost itself. It is not responsible to dispose the Player as it belongs to the user which might want to use it in another FlyleafHost. So what you need is to dispose the Player and not the FlyleafHost (or just Player.Stop() if you want to re-use it later).

I will review a FlyleafHost.Dispose() that actually closes the Surface/Overlay windows and not wait the Owner to Close().

**Comment by SuRGeoNix** [comment]
@webclient163 Just fixed a number of issues with Disposal however you should still have in your window, something like this: -

```csharp
Closed += (o, e) => Player?.Dispose();
```

You can call safely now the FlyleafHost.Dispose() independently from the main window's close

**Comment by webclient163** [comment]
Version 3.7.6, I found that all instantiated FlyleafHost must be played once before being removed from the display window for next playback, otherwise it will cause the memory to continue to rise until it is exhausted

**Comment by SuRGeoNix** [comment]
@webclient163 first of all can you please check with the 3.7.7 that I've tried to fix the dispose issues? And make sure you call Dispose for each FlyleafHost that you destroy and the same for each Player that you destroy. Ideally, a sample project would help me see what is going on and fix any issue that still remains in Flyleaf or see what you possible doing in the wrong way.

**Comment by webclient163** [comment]
My current use case is to switch between 32 or more cameras quickly, and due to the need to switch as quickly as possible, I can only turn on all the cameras and then render the required display to the specified window according to the instructions.
Keeping all cameras on consumes a lot of system resources, and switching slows down if you switch to that camera and turn it on again.
Is there a way to keep the camera connected but not decode and play, just keep the last I-frame in the buffer, and then decode and play it when switching to that camera?

**Reply by SuRGeoNix** [reply]
> Is there a way to keep the camera connected but not decode and play, just keep the last I-frame in the buffer, and then decode and play it when switching to that camera?

You could Pause the decoders and keep the demuxer running and having a task that it will destroy the demuxed packets (maybe the half of them not all - based on the Config.Demuxer.BufferDuration) so when the user clicks Player.Play() it will continue playing directly with the last cached packets. Steps:-

1) Player.Pause() [You need to ensure that it is fully paused here maybe just a thread sleep (50) for now will do the work]
2) Player.decoder.Pause() [The first step will put the pause on queue so you need an actual Pause()]
3) Player.decoder.Flush() [Cleans all the packets/frames and informs the decoder that it will have to play from a keyframe]
4) Player.VideoDemuxer.Start() [Even if the player will be paused the demuxer will run always]
5) Custom Task [Keep running a task that it will remove the Player.VideoDemuxer.VideoPackets/AudioPackets from the demuxer so the queue will not be full and you will stay at low latency]
6) Player.Play() to continue playing from the last cached packets

**Reply by webclient163** [reply]
Ok thanks
I'll try the method you provided, and I'd also like to try the way to input a custom stream to implement it.

**Comment by webclient163** [comment]
I can call FlyleafHost.Dispose() normally in version 3.7.7 to close FlyleafHost, and I need to call FlyleafHost.Player.Dispose() at the same time to close Player. The FlyleafHost instance must be played once after initialization, otherwise it will still cause memory exhaustion!

**Comment by SuRGeoNix** [comment]
@webclient163 You are right, there is a memory leak when you start the Player without having a swapchain yet (flyleafhost not rendered yet). I just identified it and I will fix it soon. Thanks for the finding :)

(It was initializing the offline renderer which is used for the extractor which had the leak)

**Comment by SuRGeoNix** [comment]
@webclient163 If you can retest with v3.7.8 just to ensure that the issue was been resolved it would be great

**Comment by webclient163** [comment]
In v3.7.8, if FlyleafHost is not presented to the window, there is still a memory leak.

**Reply by SuRGeoNix** [reply]
Can you reproduce it with the Tests/PlayerDispose project in the solution? I'm using the first button and changing the code to this:-

```csharp
private void Button_Click(object sender, RoutedEventArgs e)
{
	lock (this)
	{
		Player?.Dispose();
		Player = new Player();

		// If you want to dispose also the video view each time
		FlyleafHost?.Dispose();
		FlyleafHost = new FlyleafHost();

		//ContentControl.Content = FlyleafHost;

		// Add to test WPF control
		//FlyleafHost.Content = new Flyleaf();

		FlyleafHost.Player = Player;
		//Player.OpenAsync(SampleVideo);
	}
}
```

I'm not sure what do you mean with 'not presented so I just don't put in in the Content control (ContentControl.Content = FlyleafHost)

**Reply by webclient163** [reply]
When I call Player.OpenAsync (RtspUrl) but don't render FlyleafHost to any window, a memory leak occurs

**Reply by SuRGeoNix** [reply]
I tried to include the OpenAsync and still couldn't reproduce the mem leak. I will need a sample code for that :(

**Reply by webclient163** [reply]
```
<Window x:Class="FHJT.Media.Player.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:Media.Player"
        mc:Ignorable="d"
        KeyDown="Window_KeyDown"
        Title="DrivenSwitchingPlayer" Height="570" Width="960">
    <Grid x:Name="WindowGrid" Margin="0" Background="Black">
    </Grid>
</Window>
```

```
       public void MainWindow() {
            InitializeComponent();
            Closing += Window_Closing;
            StartEngine();
            FlyleafLib.MediaPlayer.Player player = new FlyleafLib.MediaPlayer.Player();
            player.OpenAsync("rtsp://admin:123456@172.16.10.44:554/Streaming/Channels/101");
            FlyleafHost host = new FlyleafHost();
            host.Player = player;
            WindowGrid.Children.Add(host);// Without this statement, a memory leak occurs
        }
```

**Reply by SuRGeoNix** [reply]
when the memory leak will happen? and where is your disposal code? Enabling logging might reveal the issue as well. I think I've found a possible race condition when disposing and opening. So even if you dispose it will continue to open the input which could cause a memory leak.

**Reply by SuRGeoNix** [reply]
I've fixed another one, I was able to reproduce it by delaying the Demuxer.Open which is similar to your scenario with the RTSP network stream (can have a delay). Hope this time is the right one, if not I will need a Log :)

**Reply by webclient163** [reply]
I mean don't call that player.Dispose()  and FlyleafHost.Dispose(), Memory leaks occur.

**Reply by SuRGeoNix** [reply]
So you don't talk about the dispose process that causes the leak but while the player is playing? Because you need to dispose it manually (both FlyleafHost and the Player)

**Comment by webclient163** [comment]
I have multiple monitors, and when I place a window on monitor 2 to play surveillance video, double-clicking a video will display a full-screen image of that video on monitor 1, not on monitor 2

**Reply by SuRGeoNix** [reply]
Indeed, this is an issue for the attached FlyleafHost. I will review it.

**Reply by SuRGeoNix** [reply]
I've fixed this one including some others, hopefully new FlyleafHost becomes stable enough now

**Comment by webclient163** [comment]
The overall quality of the Flyleaf player is very high and is appreciated in many ways, thanks

**Comment by webclient163** [comment]
in version 3.7.16 I have discovered a new issue:  when I play 32 way videos, the CPU usage stabilizes at around 25% in the first 5-10 minutes, but after 5-10 minutes, the CPU usage quickly increases to over 75%-90%.
This issue may be ignored when only playing single channel videos, but it becomes more serious when there are many videos being played.
CPU Monitoring:
system.private.corelib.dll!0x00007ffca717f2d7|259406 \(82.27%\)|259313ms \(82.24%\)|system.private.corelib||

**Comment by SuRGeoNix** [comment]
Do you mean that this issue was not happening in previous v3.7.15?

**Reply by SuRGeoNix** [reply]
Btw, there was an issue with 3.7.16 but I'm not sure if that was related to yours. Please provide more info to possible reproduce it?

**Reply by webclient163** [reply]
Sorry, I found out it was caused by one of the cameras being damaged
